google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+
+#include "google/cloud/pubsublite/internal/futures.h"
+#include "google/cloud/async_streaming_read_write_rpc.h"
+#include "google/cloud/internal/backoff_policy.h"
+#include "google/cloud/internal/retry_policy.h"
+#include "google/cloud/log.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <memory>
+#include <utility>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+using google::cloud::internal::BackoffPolicy;
+using google::cloud::internal::RetryPolicy;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to create new streams.
+ */
+template <typename RequestType, typename ResponseType>
+using AsyncStreamFactory = std::function<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>()>;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to initialize a stream
+ * from AsyncStreamFactory.
+ */
+template <typename RequestType, typename ResponseType>
+using StreamInitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+using AsyncSleeper = std::function<future<void>(std::chrono::milliseconds)>;
+
+using RetryPolicyFactory = std::function<std::unique_ptr<RetryPolicy>()>;
+
+/**
+ * The purpose of this class is to encapsulate the retry logic of a generic
```

Comment:
You have not answered my question about how this is used.  Without answering that question we do not know if this is the right API.  We can strip this class from this PR if you think that discussion will take a while, or we can figure out what the API should be and then merge this PR. Your call.